### PR TITLE
Navigation: Introduce a jsonData column to the preferences table to store Navbar preferences

### LINKED ISSUE #44376: Navigation: Store pinned status in a db
- [ ] decide on a structure to store in the DB. default behaviour will be to have `showInNavBar = true`
  - initial thoughts:
    - create new table, e.g. `nav_preferences`, similar to existing `preferences` structure
    - see [here](https://github.com/grafana/grafana/blob/mob/main-user-essentials/contribute/architecture/backend/database.md) for info
    - by default, nothing will be added to this table
    - rows will only be added when changes are made (e.g. something is changed to `showInNavBar = false`)
    - schema would look something like userId, orgId, navId, showInNavBar, sortWeight
    - (maybe look at library elements folder?)
- [ ] modify existing bootData to include pinned status
  - call db to get nav preferences
  - merge into existing getNavTree function
- [ ] ensure this is all behind the `newNavigation` feature toggle 

Task: 
- [ ] Transform the table migration to use the dialect migrator
- [x] Create NavBarPrerencesDTO
- [ ] Add unit tests if possible
- [x] Rename the variable `ShowInNavBar` to `HideFromNavBar`
- [x] Add sql queries to avoid messing up with the database when reviewing the PR
```
DELETE from migration_log where sql like '%navbar_preferences%';
drop table navbar_preferences;

```

For https://github.com/grafana/grafana/issues/43822

### LINKED ISSUE #44377: Navigation: Add a new API route to manage NavBar updates (pinning)
- [ ] add a new update route (POST/PUT?) to something like `/api/preferences/navbar_preferences` to update the navbar (e.g. when a new app plugin is added, when something is pinned/unpinned)
  - **structure tbd**.
  - do we want to just POST an update to `//api/preferences/navbar_preferences`?
  - do we want to PUT an individual update to `/api/preferences/navbar_preferences/{id}`?
- [ ] (maybe) add a new DELETE route to `/api/preferences/navbar_preferences/{id}` to remove an item from the navbar settings (e.g. when an app plugin is uninstalled)
  - the main question is **how do we manage uninstalling an app plugin?** we might not need a DELETE route for this
- [ ] (maybe) add a new GET API (something like `/api/preferences/navbar_preferences`?)
  - doesn't look like the frontend will use this at all, but might be useful to have the full RESTful actions
- [ ] ensure this is all behind the `newNavigation` feature toggle

For https://github.com/grafana/grafana/issues/43822

### PR DESCRIPTION
**Pull request by ashharrison90**
**What this PR does / why we need it**:

We're implementing some changes that will allow configuration of the Navbar. See https://www.figma.com/proto/esk6QuF0SSt8jY2UsggXJp/New-Navigation?page-id=1879%3A21396&node-id=1992%3A22049&scaling=min-zoom&starting-point-node-id=1992%3A22049 for the latest designs.

With that in mind, we need some way to store these customisations in the backend. The structure we've come up with so far is to add them to the existing `navTree` structure - this avoids making a GET request. Then in the frontend we port this structure to a Redux store and operate on that, pushing any updates back to the backend via a PUT request.

So these changes:
- adds a new migration to the existing preferences table to add a new `json_data` column
- adds a new api PATCH route: `/api/user/preferences`
- adds a new api PATCH route: `/api/org/preferences`

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/44376
Fixes https://github.com/grafana/grafana/issues/44377

**Special notes for your reviewer**:
We're not backend developers and not overly familiar with any code conventions we might have in the Go code, so please be extra critical! Happy to implement anything and everything if you point us in the right direction. 😅 

If you want to run locally, enable the `newNavigation` feature flag and **take a backup of your data/grafana.db**.
Once you're done testing and switch back to main, run something like:
```sql
DELETE from migration_log where migration_id like '%preferences.json_data%';
ALTER TABLE preferences DROP COLUMN json_data;
```
against your grafana.db to clear up the migrations. (or just override your grafana.db with the backup you made)

### CODE REVIEW

**Comment by ashharrison90** [comment]
@grafana/backend-platform would love some eyes on this to see if we're going in the right direction whenever anyone has a free moment. thanks! 🙏

**Comment by marefr** [comment]
@DanCech pinging you here since you introduced the key/value store. Think it could be a good fit here given that we need to store a value per org, namespace and user? The idea then would be to store a value containing all navbar preferences for a user. 

If so, we don't need to introduce another database table.

**Comment by ashharrison90** [comment]
> The idea then would be to store a value containing all navbar preferences for a user

@marefr what would the structure of that value look like then - a json object or something similar? if so that could work. was a little hesitant to do something like that as we didn't know how performant it is to store things in the db in that way. 

for extra context, there's also been discussion about setting navbar preferences at an org level (with behaviour similar to e.g. the theme preference). it's not something we're implementing now, but we wanna avoid any implementation that blocks us from doing this in the future. sounds like that could also still work in your proposal - we'd just store a separate org preference, then when checking for preferences we look for a user preference first, then an org preference if no user preference exists? i'm just rambling out loud 😅

**Comment by ashharrison90** [comment]
as long as that doesn't stop us from potentially expanding to have org level navbar preferences later, then no i don't think so.

is there an example of another service storing something in the user table i can take a look at?

**Comment by DanCech** [comment]
sounds good, I'd forgotten about the existing preferences service which does seem like a logical place to put something like this

**Comment by marefr** [comment]
Great. Also came to think about this [GetPreferencesWithDefault](https://github.com/grafana/grafana/blob/2082eeb72fbfa29c2aaa9a92ade6b4f031e1d8ef/pkg/services/sqlstore/preferences.go#L18). In case that becomes tricky to merge navbar preferences with fallbacks to org settings for example, might want/need to introduce a specific database query for navbar preferences.

In addition, introducing a new column for storing arbitrary JSON could be a good opportunity for supporting adding support for additional settings in the future without needing to add new columns to the database table.

File: pkg/models/preferences.go
Comment (DanCech) [inline pkg/models/preferences.go:53]:
```suggestion
	Navbar *simplejson.Json `json:"navbar"`
```

File: pkg/services/sqlstore/migrations/preferences_mig.go
Comment (DanCech) [inline pkg/services/sqlstore/migrations/preferences_mig.go:56]:
```suggestion
				Name: "navbar", Type: DB_Text, Nullable: true,
```

Comment (marefr) [reply pkg/services/sqlstore/migrations/preferences_mig.go:56]: This works for navbar, but considering using a more generic column that could hold potentially other future settings would be nice so that we don't need to add additional columns for each new setting

Comment (marefr) [reply pkg/services/sqlstore/migrations/preferences_mig.go:56]: Maybe `data` or `json_data`?

File: pkg/api/preferences.go
Comment (DanCech) [inline pkg/api/preferences.go:20]: Because this is a global function name within the api I'd give it a slightly more descriptive name, the same is true for the query, command and sqlstore function names.
```suggestion
func (hs *HTTPServer) SetPreferencesJsonData(c *models.ReqContext) response.Response {
```

File: pkg/api/index.go
Comment (ashharrison90) [inline pkg/api/index.go:411]: nooby go/json question: how do we type this correctly? this works, but feels wrong. the structure is something like:
```json
{
  "navbar": [{"id": "explore", "hideFromNavbar": true}]
}
```

ideally we'd define some structs, like:
```go
type NavbarPreference struct {
  Id: string `json:"id"`
  HideFromNavbar bool `json:"hideFromNavbar"`
}

type PreferencesJsonData struct {
  Navbar NavbarPreference[] `json:"navbar"`
}
```

and use those, but i can't quite figure out the syntax 🤡

**Comment by ashharrison90** [comment]
gonna continue to add some unit tests, but the implementation looks ready for a more thorough review. 😄

File: pkg/api/index.go
Comment (marefr) [reply pkg/api/index.go:411]: Then maybe it's more optimal not converting read data from database to simplejson.Json and instead convert to the structure you want/need.

But with your current approach you should be able to something like:
```go
	jsonDataBytes, err := jsonDataPref.MarshalJSON()
	if err != nil {
		return err
	}
	var v PreferencesJsonData
	if err := json.Unmarshal(jsonDataBytes, &v); err != nil {
		return err
	}
```

Comment (marefr) [reply pkg/api/index.go:411]: For the potential more optimal approach, might be able to do something like this (but haven't tested it):
```go
type Preferences struct {
	Id              int64
	OrgId           int64
	UserId          int64
	TeamId          int64
	Version         int
	HomeDashboardId int64
	Timezone        string
	WeekStart       string
	Theme           string
	Created         time.Time
	Updated         time.Time
	JsonData        *PreferencesJsonData
}

// The following needed for to implement the xorm/database ORM Conversion interface do the conversion when reading/writing to the database.

func (j *PreferencesJsonData) FromDB(data []byte) error {
	dec := json.NewDecoder(bytes.NewBuffer(data))
	dec.UseNumber()
	return dec.Decode(j)
}

func (j *PreferencesJsonData) ToDB() ([]byte, error) {
	if j == nil {
		return nil, nil
	}

	return json.Marshal(j)
}
```

**Comment by ashharrison90** [comment]
@DanCech @marefr huge thanks for the pointers! 🙌 

how does this look now? 😄

File: pkg/services/sqlstore/preferences.go
Comment (marefr) [inline pkg/services/sqlstore/preferences.go:34]: The `org_id` filter is part of every OR-clause and could be optimized. Would suggest adding that as first filter. To optimize further add a check if `UserId` is > 0 and add filter in similar was as for teams. User id could be 0 for anonymous access or use of an API key.

File: pkg/services/sqlstore/preferences.go
Comment (marefr) [inline pkg/services/sqlstore/preferences.go:115]: Would suggest reuse the `GetPreferences` method for this. Any reason not to?

File: pkg/services/sqlstore/preferences.go
Comment (marefr) [inline pkg/services/sqlstore/preferences.go:187]: Would suggest reuse the `SavePreferences` method for this. Any reason not to? It would however require reading the preferences in the API before saving in case the `JsonData` is not provided and the other way around only `JsonData` provided, but not the other preferences.

File: pkg/services/sqlstore/preferences.go
Comment (marefr) [inline pkg/services/sqlstore/preferences.go:53]: This will not merge settings, e.g. there's a default set on org and user want's to override the org preferences with only his changes. The result of that might be that org has hidden several nav items and user decide to show on of these, then all of a sudden all nav items that was hidden is now shown. Something to think about, but probably not important now.

File: pkg/api/api.go
Comment (marefr) [inline pkg/api/api.go:271]: This one doesn't feel optimal introducing a specific new endpoint for only updating navbar preferences when we already have the following:
- https://grafana.com/docs/grafana/latest/http_api/preferences/
- https://grafana.com/docs/grafana/latest/http_api/team/#get-team-preferences

No blocker for merging, but before release it would be nice to try make this a bit more consistent by
- Updating [getPreferencesFor](https://github.com/grafana/grafana/blob/10b3847d57dafd42eeb18f89a1d531fc569225cd/pkg/api/preferences.go#L40) to return JSONData preferences and/or each property in JSONData as it's own root property, similar as theme, homedashboardId etc.
- Reuse [updatePreferencesFor](https://github.com/grafana/grafana/blob/10b3847d57dafd42eeb18f89a1d531fc569225cd/pkg/api/preferences.go#L66) to allow saving JSONData preferences - if JSONData preferences not provided, don't update it to keep backward compatibility.
- Introducing a `patchPreferencesFor` method to support HTTP `PATCH` verb to only update provided preferences. With these changes I think we can deprecate `/set-home-dash` since it doesn't seem to be used by Grafana UI and then remove it completely in a major version.

**Review by marefr — CHANGES_REQUESTED** [review]
Great work! Added some comments/suggestions, but in general it looks good :+1: No big blocker for merging, but as explain in comment would be good to iterate a but on the API endpoints before release trying to keep our non-consistent API a bit more consistent :smile: 

As you might see, some comments might be something you can return to later to get this merged and keep iterating.

**Comment by ashharrison90** [comment]
@marefr think i've made the requested changes.

only one i haven't done yet is https://github.com/grafana/grafana/pull/44914#discussion_r811295463 about adding a new patch endpoint instead. ok if we do that in a follow up PR? i know that means we add this route then almost immediately deprecate it, buuuuuut would be nice to get the whole e2e feature working first before doing some refactors 😄 

lmk what you think!

**Review by marefr — COMMENTED** [review]
Sorry for slow response. Great work, starting to look really nice. Added some comments/suggestions though. Most important the `SavePreferences` in sqlstore.

> @marefr think i've made the requested changes.
> 
> only one i haven't done yet is [#44914 (comment)](https://github.com/grafana/grafana/pull/44914#discussion_r811295463) about adding a new patch endpoint instead. ok if we do that in a follow up PR? i know that means we add the `/set-json-data` route then almost immediately deprecate it, buuuuuut would be nice to get the whole e2e feature working first before doing some refactors smile
> 
> lmk what you think!

Totally fine, but #45690 might be dependent on this feature? As long as you set a plan with them to remove any double work/conflicts that it might involve when two teams working on the same feature/code and in addition have everything wrapped behind feature toggle you can go ahead and merge this (after you've addressed/responded to my comments :smile: ).

Currently no way to read navbar settings from the API `/api/user/preferences`, but guess you're planning to attack this later?

File: pkg/services/sqlstore/sqlstore.go
Comment (marefr) [inline pkg/services/sqlstore/sqlstore.go:81]: Are these changes still needed? Maybe can revert changes in this file if not needed for you?

File: pkg/api/preferences.go
Comment (marefr) [inline pkg/api/preferences.go:55]: Should be able to do this:

```suggestion
		dto.Navbar = prefsQuery.Result.JsonData.Navbar
```

File: pkg/services/sqlstore/preferences.go
Comment (marefr) [inline pkg/services/sqlstore/preferences.go:138]: Nit. I understand why you name it this way, but you already have `SavePreferences` doing an upsert so can be confusing for the consumer. This one does an insert or a patch so might make more sense to call it `PatchPreferences`. An alternative would be to add a comment like this:

```suggestion
// UpsertPreferences inserts or patches provided non-nil preferences.
func (ss *SQLStore) UpsertPreferences(ctx context.Context, cmd *models.PatchPreferencesCommand) error {
```

**Review by marefr — APPROVED** [review]
This looks really good. Great work :star2: Great that you've added tests as well :pray: Have added some minor comments/suggestions that would be good to fix before merge. Think I had an old comment regarding renaming `HideFromNavbar` to `Hide`, but that only affects you now when you've refactored so feel free to change that outside of this PR.

:heavy_check_mark:  LGTM

Things to take into consideration before releasing this without a feature toggle:
- Update [HTTP Preferences API docs](https://grafana.com/docs/grafana/latest/http_api/preferences/) 
- Update Open API definitions before releasing this without a feature toggle:
  - https://github.com/grafana/grafana/blob/main/pkg/api/docs/definitions/user_preferences.go
  - https://github.com/grafana/grafana/blob/main/pkg/api/docs/definitions/org_preferences.go

**Review by DanCech — COMMENTED** [review]
LGTM also, nice work! I would like to see @marefr 's comments addressed before merging

**Comment by ashharrison90** [comment]
have updated the preferences api docs with the patch command. haven't added documentation for the navbar structure yet as that's still in flux. i noticed we're also missing the `weekStart` key here as well. 

for updating the openapi definitions, i'll probably need a nudge in the right direction. @papagian are you best to help with that? 😄

**Comment by ashharrison90** [comment]
ok i'm referring to https://github.com/grafana/grafana/pull/40528 to do this...

had a go at adding the open api annotations with https://github.com/grafana/grafana/pull/44914/commits/7406cb2f0cdfd5971d7dfb5c496c6653bf2eb610 

however when running `make swagger-api-spec-mac` locally, it makes a lot of additional modifications to the generated docs. that **might** be correct, but i didn't wanna commit it without checking with someone first 😅

File: docs/sources/http_api/preferences.md
Comment (achatterjee-grafana) [inline docs/sources/http_api/preferences.md:70]: Can we add a sentence - an intro/overview. It is OK if it is brief.

File: docs/sources/http_api/preferences.md
Comment (achatterjee-grafana) [inline docs/sources/http_api/preferences.md:146]: Can we add a sentence - an intro/overview. It is OK if it is brief

**Review by achatterjee-grafana — COMMENTED** [review]
Added doc request.

File: docs/sources/http_api/preferences.md
Comment (ashharrison90) [reply docs/sources/http_api/preferences.md:70]: i've added a short description, but it's worth mentioning that neither the `GET` or `PUT` requests currently have a description either, so `PATCH` is now the odd one out 😄

File: pkg/api/dtos/prefs.go
Comment (papagian) [reply pkg/api/dtos/prefs.go]: If you mean guidelines for updating the Open API definitions, yes.

**Comment by papagian** [comment]
> have updated the preferences api docs with the patch command. haven't added documentation for the navbar structure yet as that's still in flux. i noticed we're also missing the `weekStart` key here as well.
> 
> for updating the openapi definitions, i'll probably need a nudge in the right direction. @papagian are you best to help with that? 😄

You have to update [this](https://github.com/grafana/grafana/blob/main/pkg/api/docs/definitions/user_preferences.go) and [this](https://github.com/grafana/grafana/blob/main/pkg/api/docs/definitions/org_preferences.gohttps://github.com/grafana/grafana/blob/main/pkg/api/docs/definitions/org_preferences.go) files and add that PATCH swagger route. 
Since this route expects a body parameter you need define one (similar to the [UpdateUserPreferencesParam](https://github.com/grafana/grafana/blob/9aa4bde7fc9ccc9faeb8904d14ca753478546a7c/pkg/api/docs/definitions/user_preferences.go#L26))
Make sure that in `swagger:parameters` annotation you have add both tags of your PATCH routes.

Then you have to run (with enterprise enabled  so that enterprise routes are included in the generated spec):
`make swagger-api-spec`

You can inspect the generated specification by running:
`GF_FEATURE_TOGGLES_ENABLE=swaggerUi make run`
and navigate to `http://localhost:3000/swagger-ui`

If the routes are as expected then you can commit the updated `api-spec.json` and `api-merged.json`

File: public/api-merged.json
Comment (ashharrison90) [inline public/api-merged.json:14494]: these changes are unrelated to ours, but seem to be included when running `make swagger-api-spec-mac`

Comment (papagian) [reply public/api-merged.json:14494]: This change in the model was introduced [here](https://github.com/grafana/grafana-enterprise/pull/2989/files#diff-4a6a200bd703d6cdc5258455569746f486c44c861ced31ea423abc18b41cb3ba) but the specification was not updated then. It's fine to keep it.

File: public/api-merged.json
Comment (papagian) [inline public/api-merged.json:14965]: same as above: this change in the model was introduced [here](https://github.com/grafana/grafana-enterprise/pull/2989/files#diff-4a6a200bd703d6cdc5258455569746f486c44c861ced31ea423abc18b41cb3ba)

**Review by papagian — APPROVED** [review]
The HTTP API specification changes LGTM

**Comment by ashharrison90** [comment]
@papagian merged with main and reran the api spec generation, there are now **a lot** of changes. i think https://github.com/grafana/grafana/pull/46548 was run without enterprise enabled which has removed all the enterprise routes? can you confirm if that looks correct? 🤔

**Comment by ashharrison90** [comment]
@marefr i'm going to merge this unless there's any other changes you think are needed. speak now! 😄

**Comment by marefr** [comment]
Already approved and haven't looked at the latest changes. Go for it :boom:

**Comment by papagian** [comment]
> @papagian merged with main and reran the api spec generation, there are now **a lot** of changes. i think #46548 was run without enterprise enabled which has removed all the enterprise routes? can you confirm if that looks correct? 🤔

you are right! thank you for noticing and taking care of it.

**Pull request merged**
